### DISCUSSION TITLE
Candidate Backlinks: Feedback on the preview feature

### DISCUSSION THREAD

**Discussion by ScriptAutomate** [dendronhq/dendron#3583] [RFCs/Ideas]
> As a reminder, [Candidate Backlinks](https://wiki.dendron.so/notes/yxkn87ohgomk0tgs12dppur/#explicit-and-candidate-backlinks) are a **preview feature** so there are still many rough edges.

## Description

I was looking at some issues after trying out [Candidate Backlinks](https://wiki.dendron.so/notes/yxkn87ohgomk0tgs12dppur/#explicit-and-candidate-backlinks) today:
- #915
- #1059

I think the functionality currently provided by candidate backlinks could be revisited. I'm not sure if it would make sense for this feedback to become a GitHub Discussion, or potentially an RFC (not a new feature, but quite an expansion of an existing feature), or a series of different feature enhancement requests. So, I created this discussion.

## Candidate backlinks definition

> **Explicit and Candidate Backlinks**
>
> Dendron supports both explicit (wiklilinks, note refs) and candidate backlinks.
> 
> A note has a candidate backlink if a different note has the name of the current note appear anywhere in its text. For example, if your current note is named apple, and in a different note called recipes, the word apple appears in the note body but is not explicitly linked, then recipes is said to have a link candidate to the apple note.
> 
> This feature is currently disabled by default. In order to enable link candidates, add the following configuration in dendron.yml:
> ```yml
> dev:
>   enableLinkCandidates: true
> ```
> There is an icon next to each reference to indicate what kind of backlink it is. Explicit links have a link/chain icon. Candidates have the double plug / unplugged icon. See the sections below for more information on each item.
> 
> If you click on an item anywhere other than the arrow, it will open that note in your editor. Furthermore, clicking on a link candidate item will open the note that contains the candidate, and convert the link candidate into a wikilink, turning it into Linked backlink.

## Thoughts on current state

Candidate backlinks only seems to work if you use the exact full filename (`this.is.my-note.over-here`), with the dot notation included.

Looks like I can get it to work, but only because it is implemented in a very unintuitive/unexpected way:

- You can see now what is meant by the plug icon vs. the link/chain icon (in screenshot). The plug there is referencing a candidate backlink.

![image](https://user-images.githubusercontent.com/5951023/192071447-7a39bb27-0b7b-4012-ac9d-12c406af6a27.png)

- The only way this works is if there is a match with the filename itself as opposed to either the the `title` or the `basename` (as mentioned in #1059) (`basename` being `python` in `lang.python`, I think). I've added a comment to that issue around basename, because it brought up questions for me there, too.
- This means my candidate link only worked because of typing `Testing out templates.derek.my-note right now` in order for it to appear, so to me this seems like not the most helpful feature at the moment since people would rarely use that notation in the middle of a note (unless that feature is for people who are hesitant to link to an existing note, or hesitant to link to an existing note and just want a placeholder? But I see that as being a really rare situation/workflow to support).
- Candidate Backlinks will only appear in the UI under the target note itself, as opposed to any visual marker in the note you are in that has a candidate backlink in the text. As in: you can only see candidate backlinks that may be _pointing to_ your current note in the Backlinks view. You would only be aware that your current note may link out to a different note with a candidate backlink if you open the local Graph Panel (which would show connections between both notes as though it were a general backlink), but the links shown in the panel are otherwise indistinguishable from regular backlinks (no hints that there may be an opportunity for a word/phrase/series of words to become a backlink).
- Auto-creating a backlink by navigating to a note containing a candidate backlink doesn't seem to convert the link into a backlink as stated in:
  > Furthermore, clicking on a link candidate item will open the note that contains the candidate, and convert the link candidate into a wikilink, turning it into Linked backlink.

## How to test out candidate backlinks

- Follow the directions in **Candidate Backlinks Definition** above to edit your `dendron.yml`
- Reload your workspace
- Make `my.new.note`
- Make `my.new.other-note`
- Inside of `my.new.other-note`, type: `I'm looking at my.new.note today`
- Go to `my.new.note`
- On the Dendron side bar, open the Backlinks drop-down menu
- You will see that `my.new.other-note` is listed as a backlink, and if you collapse it, it will show the plug icon
  - If you don't see if, you may need to `Dendron: Reload Index` for it to appear.
  - If it still doesn't appear, it may be something having to do with how candidate backlinks aren't properly triggering an index. They seem to appear in backlink view if an explicit backlink is present in the same file.
- It also looks like this is only _current vault context wide_ (as stated in the comments of this issue)

## What does an ideal workflow look like?

I feel like ideal scenario would be:
- Multi-vault support (perhaps config optional?)
- Some way to enable hints in the UI for when candidate backlinks exist in the content of the current note (which may be the main focus of the feature request in #915)
- Some way to enable a sensitivity filter to scope when it comes to naming convention it seeks to match (matching `title` value of a note, for example, maybe?)
- Fix auto-creating of a backlink when clicking on the candidate backlink, if it is meant to function that way (I could be interpreting the steps incorrectly)

I'm not sure what the demand for this feature is, or would be, or if users would want/expect candidate backlinks to function differently. It seems good, in theory, to become aware of potentially hidden connections across notes. It sounds especially cool if being able to do so in multi-vault scenarios, as a note I've made may now have a candidate link to newly imported notes.

**Comment by rlh1994** [comment]
I think something else that would be useful for this and provide a more powerful experience would be a way to review all suggested candidate back links rather than just having them all enabled/disabled, perhaps via a doctor command.

The idea being that this would help users discover hidden connections in their notes that either they didn't know about, or just that as the vault grew there was the potential for linking to older notes.

**Comment by jeep** [comment]
For me this feature only work sporadically, at best. Reload index isn't enough, but creating a hard link will cause the candidate links on that page to show up. After that it works even if I remove the hard link. 

And then when I follow a candidate link, the wrong part is "linked" for example, In the sample above, when I follow the candidate link to my.new.note from the backlinks, This is what I get: `I'm looking at [[my.new]].note today`
 I expected `I'm looking at [[my.new.note]] today`. @ScriptAutomate seems to indicate that even that doesn't work for him.
 
And yes, it would be much better if each "node" of the name was considered independent. Ex. I use "python" then "reference.languages.python" would have a candidate backlink to the note where I typed python. If I typed "I like programing languages" in note1, then reference.language would have a candidate backlink to note1. Also, I should match title or a node in the filename.

Unlinked references can be slow in org-roam, but I never had any complaints with the performance. I used the feature a lot when I decided I wanted to split notes... I would split the note, use the unlinked references to find places where the concept was mentioned to create real links.

**Comment by kevinslin** [comment]
thanks everyone for providing feedback. as a reminder, candidate backlinks are a preview feature so there are still many rough edges.
we haven't had bandwidth to update this feature in quite a while (and actually broke it with recent refactoring).

some context: we're currently in the middle of switching our metadata backend from a json blob to sqlite. this should provide near instaneous (<100ms) latencies for all operations in dendron. performance has been a major concern for the candidate backlink feature which is why we've been holding out on working on this until we've finished the migration

in terms of what we want the functionality to be:

- currently
  - fix issues in loading workspace when enableCandidates is enabled in large vaults

- phase I
  - fuzzy match terms with potential link candidates based on the last part of a hierarchy (eg. `fruits.apples` should match `apples`)
    - should also match [alias notes](https://github.com/dendronhq/dendron/issues/128)
  - doctor command to see all link candidates in a vault as mentioned by @rlh1994

- phase II
  - highlight link candidates in the UI 
  - create note, vault, and workspace specific filters on what can be a link candidate

**Reply by ScriptAutomate** [reply]
This all sounds awesome and makes sense for the dependency on the metadata backend for performance concerns.

I've now updated the discussion title, and added a note in the initial post, that includes the point about this being a preview feature.
